### PR TITLE
Issue 2144 - Proven Headers - Protocol Version Enum.

### PR DESCRIPTION
**Pull request by ferdeen**
Related to #2144.

This is the first PR, of two, which will implement an enumeration class.  

This change is required so we allow Bitcoin and Stratis versions to become independent of each other going forward.  Also to be used as part of the proven headers work.

The code pattern was taken from https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types.

The next PR will tackle replacing the original `ProtocolVersion` enum, in the StratisBitCoin projects, with the new class (and any related tests): 

image

### CODE REVIEW

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+        /// <summary>
+        /// BIP 0031, pong message, is enabled for all versions AFTER this one.
+        /// </summary>
+        public static ProtocolVersion Bip31 = new ProtocolVersion(60000, nameof(Bip31).ToLowerInvariant());
+
+        /// <summary>
+        /// "mempool" command, enhanced "getdata" behaviour starts with this version.
+        /// </summary>
+        public static ProtocolVersion MempollGetData = new ProtocolVersion(60002, nameof(MempollGetData).ToLowerInvariant());
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: Typo, also elsewhere where the name is used

**Comment by Aprogiena** [comment]
I'm not sure what you're going to do next, so I'm not sure it is a good idea to have this separated into two PRS.

Also when you copy and paste code from somewhere else, you must make sure you change it according to our policies. Often also you should remove all the functionality that you don't want. In the comment you should always link the original source from where you inspired yourself

**Comment by ferdeen** [comment]
@Aprogiena I'll remove the unused methods from `Enumeration` class.  I've checked the source of the class and this comes under the MIT license.

I'll reference the source in the code that is using this class.

With regards to the two PR's.  I wanted to discuss the first PR to make sure this is the best approach.  

The second PR doesn't exist yet.  It's on this branch [ferdeen/StratisBitcoinFullNode@2114_proven_headers_protocol_version_enum...ferdeen:pr2_replace_enum_with_class_enum](url)).  There are quite a few changes.

Before we create the second PR, or include the code changes to the first one we can have a call on the approach.

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
+﻿using System.Collections.Generic;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Networks
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: namespace not corresponding to file location

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    /// <para>
+    /// Future versions will go into either Bitcoin <see cref="BitcoinProtocolVersion"/> or Stratis <see cref="StratisProtocolVersion"/>.
+    /// </para>
+    /// </remarks>
+    public class ProtocolVersion : Enumeration
+    {
+        /// <summary>
+        /// Protocol version.
+        /// </summary>
+        public static ProtocolVersion Protocol = new ProtocolVersion(70012, nameof(ProtocolVersion).ToLowerInvariant());
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: all of those can be readonly

File: src/NBitcoin/Enumeration.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+
+namespace NBitcoin
+{
+    /// <summary>
+    /// Enumeration classes that enable all the rich features of an object-oriented language
+    /// <seealso cref="https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types"/>
+    /// </summary
```
Comment (noescape00) [inline src/NBitcoin/Enumeration.cs]: xmltag not closed

File: src/NBitcoin/Enumeration.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public override string ToString() => this.Name;
+
+        public override bool Equals(object obj)
+        {
+            if (!(obj is Enumeration otherValue))
+                return false;
+
+            var typeMatches = this.GetType().Equals(obj.GetType());
```
Comment (noescape00) [inline src/NBitcoin/Enumeration.cs]: unvar here and on next line

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    /// <summary>
+    /// Class enumeration containing protocol versions, by name and version number (Id).
+    /// </summary>
+    /// <remarks>
+    /// This class contains baseline versions.
+    /// <para>
+    /// Future versions will go into either Bitcoin <see cref="BitcoinProtocolVersion"/> or Stratis <see cref="StratisProtocolVersion"/>.
+    /// </para>
+    /// </remarks>
+    public class ProtocolVersion : Enumeration
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: Probably rename to BaseProtocolVercion or smth and mention in the comments that those are common between stratis and bitcoin

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+        /// <summary>
+        /// ! Version after which witness support potentially exists.
+        /// </summary>
+        public static ProtocolVersion Witness = new ProtocolVersion(70012, nameof(Witness).ToLowerInvariant());
+
+        /// <summary>
+        /// short-id-based block download starts with this version.
+        /// </summary>
+        public static ProtocolVersion ShortIdBlocks = new ProtocolVersion(70014, nameof(ShortIdBlocks).ToLowerInvariant());
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: it's not for base, it's btc only

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/BitcoinProtocolVersion.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Stratis.Bitcoin.Networks
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/BitcoinProtocolVersion.cs]: wrong namespace

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/StratisProtocolVersion.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        public StratisProtocolVersion() { }
+
+        protected StratisProtocolVersion(int id, string name) : base(id, name) { }
+
+        /// <summary>
+        /// List all baseline, plus new Stratis protocol versions.
+        /// </summary>
+        public override IEnumerable<ProtocolVersion> List()
+        {
+            var list = base.List().ToList();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/StratisProtocolVersion.cs]: unvar;
why List().ToList()? Maybe change base class to return you a List<> type?

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/StratisProtocolVersion.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Networks
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/StratisProtocolVersion.cs]: wrong namespace

File: src/NBitcoin/Enumeration.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace NBitcoin
+{
+    /// <summary>
+    /// Enumeration classes that enable all the rich features of an object-oriented language
+    /// <seealso cref="https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types"/>
+    /// </summary
+    public abstract class Enumeration : IComparable
+    {
+        public string Name { get; private set; }
```
Comment (Aprogiena) [inline src/NBitcoin/Enumeration.cs]: Why would we need the name?

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        /// <summary>
+        /// Protocol version.
+        /// </summary>
+        public static ProtocolVersion Protocol = new ProtocolVersion(70012, nameof(ProtocolVersion).ToLowerInvariant());
+
+        /// <summary>
+        /// Alternate coin protocol version.
+        /// </summary>
+        public static ProtocolVersion AltProtocal = new ProtocolVersion(70000, nameof(AltProtocal).ToLowerInvariant());
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: typo

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs (hunk 2 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: ok

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs (hunk 3 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: ok

File: src/NBitcoin/Enumeration.cs (hunk 4 above)
Comment (ferdeen) [reply src/NBitcoin/Enumeration.cs]: ok

File: src/NBitcoin/Enumeration.cs (hunk 5 above)
Comment (ferdeen) [reply src/NBitcoin/Enumeration.cs]: ok

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs (hunk 6 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: Yes, I did have this as `ProtocolVersionBase` originally - I was waiting for feedback.  Cool, I'll rename and add comments accordingly.

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs (hunk 7 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: ok

**Comment by Aprogiena** [comment]
> The second PR doesn't exist yet. It's on this branch (ferdeen/StratisBitcoinFullNode@2114_proven_headers_protocol_version_enum...ferdeen:pr2_replace_enum_with_class_enum). There are quite a few changes.

I have looked into that, I am not sure about the approach. @dangershony Can you please check that?

File: src/NBitcoin/Enumeration.cs (hunk 11 above)
Comment (noescape00) [reply src/NBitcoin/Enumeration.cs]: We don't. It seems the solution was copypasted but not yet adapted for our needs

File: src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+        protected ProtocolVersion(int id, string name) : base(id, name) {}
+
+        /// <summary>
+        /// List all baseline versions.
+        /// </summary>
+        public virtual IEnumerable<ProtocolVersion> List() =>
+            new[] {
+                Protocol,
+                AltProtocal,
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Networks/ProtocolVersion/ProtocolVersion.cs]: The idea is to get rid of this enum `AltProtocal`

**Comment by ferdeen** [comment]
Protocol version is going to be postponed

We will reuse the Bitcoin 70012 version on stratis to signal proven headers.

For reference this was the current work done so far - 

- master (https://github.com/stratisproject/StratisBitcoinFullNode/commit/df16fb06366202bd8e49eb55e16901e0540a7535 17/09/2018) -> branch 1  https://github.com/stratisproject/StratisBitcoinFullNode/compare/master...ferdeen:2114_proven_headers_protocol_version_enum

- branch 1 -> branch 2  https://github.com/ferdeen/StratisBitcoinFullNode/compare/2114_proven_headers_protocol_version_enum...ferdeen:pr2_replace_enum_with_class_enum#diff-072ecedd2167547d401563c7f0a76c3a

Issue #2144 mentions using a mapper type data structure as part of the solution.  Which will be looked at when the task is picked up at a later date.

**Pull request closed without merging**